### PR TITLE
Make multiple fixes until it works

### PR DESCRIPTION
**Pull request by PeganovAnton**
1. Add logits to `LatentVariableInferenceModel.forward` in order to                                                     
   perform validation.                                                                                                  
Pytorch Ignite Engine for evaluation demands and inference function which will return logits.                           
                                                                                                                        
2. Refactor `train()` function.                                                                                         
                                                                                                                        
3. Fix perplexity computation. Perplexity is not exponent of loss (cross entropy) but an                                
   exponent of entropy.                                                                                                 
Create a pytorch ignite metric                                                                                          
                                                                                                                        
4. Replace `RandomSampler` with `DistributedSampler` for distributed data parallel case                                 
`DistrubutedSampler` splits selects only a part of data whereas `RandomSampler` loads all                               
dataset. If `RandomSampler` is used 1 epoch actually consists of nproc epochs. And there                                
can be repeated elements in a batch.                                                                                    
                                                                                                                        
5. Require at least one of options `--do_train`, `--do_eval`                                                            
                                                                                                                        
6. Do not create objects if they are not used. For instance, if no evaluation is                                           
performed, `evaluator` is not created.                                                                                  
                                                                                                                        
7. Do evaluation in parallel if distributed data parallel mode is used.                                                 
                                                                                                                        
6. Make batch a dictionary                                                                                              
It is difficult to maintain the order of model inputs in the dataset, collate function                                  
and train function. So now the batch is a dictionary.                                                                   
                                                                                                                           
7. Remove `history_folded` entirely since it is not used.                                                                
                                                                                                                           
8. Make `collate_dialog()` function a dataset method.                                                                      
This allows to make `num_candidates` an attribute and remove it from samples and batch.                                    
                                                                                                                           
9. The elements of `PersonaChatDataset.dataset[key]` where key is one of `'input_ids'`,                                      
   `'token_type_ids'`, `'mc_token_ids'`, `'lm_labels'`, `'mc_labels'` are now ready samples.                                       
   No slicing is needed to get an element.                                                                                 
                                                                                                                           
10. Remove max_num_persona. The number of personas in a batch can vary now.                                                
Previously there always were max_num_persona personas in a batch regardless of                                             
`num_beams`. That lead to high GPU memory usage and slowed down computations.                                              
                                                                                                                           
11. Fix `num_labels` local variable in `LatentVariableInferenceModel.forward()`.                                           
Previously it became equal to zero if `--num_candidates` was greater than 1.

### CODE REVIEW

**Comment by majumderb** [comment]
Thanks for cleaning this up! Looks good to me!
